components/region_cache_memory_engine/src/background.rs

```diff
@@ -451,6 +618,76 @@ impl BackgroundRunnerCore {
         #[cfg(test)]
         flush_epoch();
     }
+
+    fn evict_on_soft_limit_reached(&mut self) {
+        if self.range_stats_manager.is_none() {
+            warn!("range stats manager is not initialized, cannot evict on soft limit reached");
+            return;
+        }
+        let range_stats_manager = self.range_stats_manager.as_ref().unwrap();
+        let to_shrink_by = self
+            .memory_controller
+            .mem_usage()
+            .checked_sub(self.memory_controller.soft_limit_threshold());
+        if to_shrink_by.is_none() {
+            return;
+        }
+        let mut remaining = to_shrink_by.unwrap();
+        let mut ranges_to_evict = Vec::<(CacheRange, u64)>::with_capacity(256);
+        range_stats_manager.collect_candidates_for_eviction(&mut ranges_to_evict);
+        let ranges_to_evict = ranges_to_evict
+            .iter()
+            .filter(|(range, _)| self.engine.read().range_manager().contains_range(range))
+            .rev();
+        for (range, approx_size) in ranges_to_evict {
+            if remaining == 0 {
+                break;
+            }
+            if self.engine.write().mut_range_manager().evict_range(range) {
+                info!("evict on soft limit reached"; "range" => ?&range, "approx_size" => approx_size, "remaining" => remaining);
+                remaining = remaining
```

Comment:
There maybe a huge difference between the memory usage of the region in in-memory engine and the region approximate size.

Reply:
Yes, that is true. I understand this is imprecise, so there is a best effort approaches.  Approximate size is via region info. Do we have a preferred way to check this that is more direct?

Reply:
@SpadeA-Tang Looks like there's no clean API to get memory usage of a range directly from the range cache engine. I do have an idea on this, but it might take some time to implement. On the other hand, approximate size is most likely to be smaller than the actual size - which would lead to over eviction, which seems safer than under eviction. Any thoughts?

Reply:
It's indeed over eviction. Actually, the approximate region size is generally estimazted by including all MVCC versions as well as rocksdb tombstones where there's may be only a small fraction of them that is in the memory engine. So this difference may be significantly huge.

Reply:
@SpadeA-Tang How difficult would it to be to add an API for accurate size? Seems like adding a quick lock free way to keep track of a region (perhaps by using thread local to set region id (or start and end keys) when allocating memory for that region, or keeping track of all allocations for a specific range some other way?) would be the right way to do so. I can create an issue for this.

Reply:
@SpadeA-Tang What do you think about making delete_range return the number of bytes it has deleted? I can make this change in this PR. It might have slight overhead. I don't see any drawback to that.